burn-core/src/nn/lstm/lstm.rs

```diff
@@ -0,0 +1,303 @@
+        let mut batched_hidden_state = Tensor::zeros([self.batch_size, seq_length, self.d_hidden]);
+
+        let (mut cell_state, mut hidden_state) = match state {
+            Some((cell_state, hidden_state)) => (cell_state, hidden_state),
+            None => (
+                Tensor::zeros([self.batch_size, self.d_hidden]),
+                Tensor::zeros([self.batch_size, self.d_hidden]),
+            ),
+        };
+
+        for t in 0..seq_length {
+            let indices = Tensor::arange(t..t + 1);
+            let input_t = batched_input.clone().index_select(1, indices).squeeze(1);
+            // f(orget)g(ate) tensors
+            let biased_fg_input_sum = self.gate_product(&input_t, &hidden_state, &self.forget_gate);
+            let forget_values = activation::sigmoid(biased_fg_input_sum); // to multiply with cell state
+
+            // i(nput)g(ate) tensors
+            let biased_ig_input_sum = self.gate_product(&input_t, &hidden_state, &self.input_gate);
+            let add_values = activation::sigmoid(biased_ig_input_sum);
+
+            // o(utput)g(ate) tensors
+            let biased_og_input_sum = self.gate_product(&input_t, &hidden_state, &self.output_gate);
+            let output_values = activation::sigmoid(biased_og_input_sum);
+
+            // c(ell)g(ate) tensors
+            let biased_cg_input_sum = self.gate_product(&input_t, &hidden_state, &self.cell_gate);
+            let candidate_cell_values = biased_cg_input_sum.tanh();
+
+            cell_state = forget_values * cell_state.clone() + add_values * candidate_cell_values;
+            hidden_state = output_values * cell_state.clone().tanh();
+
+            // store the state for this timestep
+            batched_cell_state = batched_cell_state.index_assign(
+                [0..self.batch_size, t..(t + 1), 0..self.d_hidden],
+                cell_state.clone().unsqueeze(),
+            );
+            batched_hidden_state = batched_hidden_state.index_assign(
+                [0..self.batch_size, t..(t + 1), 0..self.d_hidden],
+                hidden_state.clone().unsqueeze(),
+            );
+        }
+
+        (batched_cell_state, batched_hidden_state)
+    }
+
+    /// Helper function for performing weighted matrix product for a gate and adds
+    /// bias, if any.
+    ///
+    ///  Mathematically, performs `Wx*X + Wh*H + b`, where:
+    ///     Wx = weight matrix for the connection to input vector X
+    ///     Wh = weight matrix for the connection to hidden state H
+    ///     X = input vector
+    ///     H = hidden state
+    ///     b = bias terms
+    fn gate_product(
+        &self,
+        input: &Tensor<B, 2>,
+        hidden: &Tensor<B, 2>,
+        gate: &GateController<B>,
+    ) -> Tensor<B, 2> {
+        let input_product = input.clone().matmul(gate.get_input_weight());
+        let hidden_product = hidden.clone().matmul(gate.get_hidden_weight());
+
+        match (gate.get_input_bias(), gate.get_hidden_bias()) {
```

Comment:
Using `pub(crate)` to access `input_transforms`(GateController) and `bias`(Linear)

```rust
let input_bias = gate.input_transforms.bias.map(Param::val);
```

Reply:
@nathanielsimard I understand the case against getters, and I'm not going to push the issue, but I think that one could argue for readability/understandability here. To make the above work, I needed to do this:

```rust
let input_bias = gate
            .input_transform
            .bias
            .as_ref()
            .map(|bias_param| bias_param.val());
```
Direct field access is certainly more concise and cuts down on verboseness, but it's also less clear what is happening unless you are familiar with the structure of the `GateController` and `Linear` structs, and how they work with `Param`. Just food for thought.